ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite_core.cc

```diff
@@ -437,68 +384,310 @@ TFLiteCore::setInputTensorDim (const GstTensorsInfo * info)
+ */
+int
+TFLiteCore::setOutputTensorProp ()
+{
+  int err;
+
+  interpreter.lock();
+  err = interpreter.setOutputTensorProp ();
+  interpreter.unlock();
+
+  return err;
+}
+
+/**
+ * @brief	return the Dimension of Input Tensor.
+ * @param[out] info Structure for tensor info.
+ * @todo return whole array rather than index 0
+ * @return 0 if OK. non-zero if error.
+ */
+int
+TFLiteCore::getInputTensorDim (GstTensorsInfo * info)
+{
+  interpreter.lock ();
+  gst_tensors_info_copy (info, interpreter.getInputTensorsInfo());
+  interpreter.unlock ();
+
+  return 0;
+}
+
+/**
+ * @brief	return the Dimension of Tensor.
+ * @param[out] info Structure for tensor info.
+ * @todo return whole array rather than index 0
+ * @return 0 if OK. non-zero if error.
+ */
+int
+TFLiteCore::getOutputTensorDim (GstTensorsInfo * info)
+{
+  interpreter.lock ();
+  gst_tensors_info_copy (info, interpreter.getOutputTensorsInfo());
+  interpreter.unlock ();
 
   return 0;
 }
 
+/**
+ * @brief set the Dimension for Input Tensor.
+ * @param info Structure for input tensor info.
+ * @return 0 if OK. non-zero if error.
+ * @note rank can be changed dependent on the model
+ */
+int
+TFLiteCore::setInputTensorDim (const GstTensorsInfo * info)
+{
+  int err;
+
+  interpreter.lock ();
+  err = interpreter.setInputTensorsInfo (info);
+  interpreter.unlock ();
+
+  return err;
+}
+
+/**
+ * @brief	reload a model
+ * @param	tflite	: the class object
+ * @param[in] model_path : the path of model file
+ * @return 0 if OK. non-zero if error.
+ * @note reloadModel() is asynchronously called with other callbacks
+ */
+int
+TFLiteCore::reloadModel (const char * _model_path)
+{
+  int err;
+
+  interpreter_sub.lock();
+
+  /**
+   * load a model into sub interpreter. This loading overhead is indenendent
+   * with main one's activities.
+   */
+  err = interpreter_sub.loadModel (_model_path, use_nnapi);
+  if (err != 0) {
+    g_critical ("Failed to load model %s\n", _model_path);
+    goto out_unlock;
+  }
+  err = interpreter_sub.setInputTensorProp ();
+  if (err != 0) {
+    g_critical ("Failed to initialize input tensor\n");
+    goto out_unlock;
+  }
+  err = interpreter_sub.setOutputTensorProp ();
+  if (err != 0) {
+    g_critical ("Failed to initialize output tensor\n");
+    goto out_unlock;
+  }
+
+  /* Also, we need to check input/output tensors have the same info */
+  if (!gst_tensors_info_is_equal (
+        interpreter.getInputTensorsInfo(),
+        interpreter_sub.getInputTensorsInfo()) ||
+      !gst_tensors_info_is_equal (
+        interpreter.getOutputTensorsInfo(),
+        interpreter_sub.getOutputTensorsInfo())) {
+    g_critical ("The model has unmatched tensors info\n");
+    err = -EINVAL;
+    goto out_unlock;
+  }
+
+  /**
+   * Everything is ready. let's move the model in sub interpreter to main one.
+   * But, it needs to wait if main interpreter is busy (e.g., invoke()).
+   */
+  interpreter.lock();
+  interpreter.moveInternals (interpreter_sub);
```

Comment:
This style of loading multiple interpreter blocks the main interpreter for shorter duration.
However, this causes double memory requirement because both the models are loaded into memory at this moment.
As nnstreamer targets embedded devices with limited resources, I am not sure if this is better (can cause of ENOMEM) or unloading main interpreter first and then loading the new model.

Reply:
Yes, I think each impl. has a tradeoff. But, I think the latter one (i.e., unloading old model and loading new model) has additional disadvantages. If a new model has an incompatible hierarchy with the existing pipeline, loadModel() is failed and then we need to reload old model again. That's a quite big overhead.

So, I think it's better to limit this feature by only enabling reload() in devices with enough memory size.

Reply:
Then, lets add a note that this function needs twice the memory of a single model.

Reply:
Yeap. Good suggestion.